pubnub-java/src/integrationTest/java/com/pubnub/api/integration/PresenceIntegrationTests.java

```diff
@@ -0,0 +1,388 @@
+    @Test
+    public void should_setState_withHeartbeat() throws InterruptedException {
+//        enableHeartbeatLoop(2);
+        boolean WITH_HEARTBEAT_TRUE = true;
+        final AtomicInteger hits = new AtomicInteger();
+        final int expectedHits = 2;
+
+        final JsonObject expectedStatePayload = generatePayload();
+        final String expectedChannel = RandomGenerator.get();
+
+        pubNub.addListener(new SubscribeCallback.BaseSubscribeCallback() {
+
+            @Override
+            public void presence(@NotNull PubNub pubnub, @NotNull PNPresenceEventResult presence) {
+                System.out.println("---" + presence.getEvent());
+                if (presence.getEvent().equals(STATE_CHANGE_EVENT)
+                        && presence.getChannel().equals(expectedChannel)
+                        && presence.getUuid().equals(pubNub.getConfiguration().getUserId().getValue())) {
+                    assertEquals(expectedStatePayload, presence.getState());
+                    hits.incrementAndGet();
+                }
+            }
+        });
+
+        subscribeToChannel(pubNub, expectedChannel);
+
+        pubNub.setPresenceState()
+                .channels(Collections.singletonList(expectedChannel))
+                .state(expectedStatePayload)
+                .withHeartbeat(WITH_HEARTBEAT_TRUE)
+                .async((result) -> {
+                    assertFalse(result.isFailure());
+                    assertEquals(expectedStatePayload, result.getOrNull().getState());
+                });
+
+//        Awaitility.await().atMost(Durations.FIVE_SECONDS).untilAtomic(hits, IsEqual.equalTo(1));
+        Thread.sleep(1000);
+
+        pubNub.getPresenceState()
+                .channels(Collections.singletonList(expectedChannel))
+                .async((result) -> {
+                    assertFalse(result.isFailure());
+                    assertEquals(expectedStatePayload.get("text"), result.getOrNull().getStateByUUID().get(expectedChannel).getAsJsonObject().get("text"));
+                    assertEquals(expectedStatePayload.get("info"), result.getOrNull().getStateByUUID().get(expectedChannel).getAsJsonObject().get("info"));
+                    assertEquals(expectedStatePayload.get("uncd"), result.getOrNull().getStateByUUID().get(expectedChannel).getAsJsonObject().get("uncd"));
+                    hits.incrementAndGet();
+                });
+
+//        Awaitility.await().atMost(Durations.FIVE_SECONDS).untilAtomic(hits, IsEqual.equalTo(expectedHits));
+        Thread.sleep(1000);
+    }
+
+    @Test
+    public void testPresenceState() {
+        final AtomicInteger hits = new AtomicInteger();
+        final int expectedHits = 2;
+
+        final JsonObject expectedStatePayload = generatePayload();
+        final String expectedChannel = RandomGenerator.get();
+
+        pubNub.addListener(new SubscribeCallback.BaseSubscribeCallback() {
```

Comment:
SubscribeCallback.BaseSubscribeCallback shouldn't really be used anymore, it's only here for backward compatibility to not break existing code (such as this test)

in a normal situation, this should be just `new EventListener { override fun presence ... }`
similarly, when adding a status listener it would be just `new StatusListener { override fun status ... }`

Reply:
and if you want both status AND events, you can still just do `new SubscribeCallback { ... }`